feat: support importing R data files into JASP

shun2wang opened the pull request:
Now we using [librdata](https://github.com/WizardMac/librdata) to import .RData/.rds files into JASP! Compiling librdata on Windows is still not silky🤔 because librdata is not on conan but I had opened a PR for conan repo, but it cannot build with msvc...see:https://github.com/conan-io/conan-center-index/pull/26024

Close: https://github.com/jasp-stats/jasp-issues/issues/2884

TODO and Help wanted:

- [x] Make it work on macOS
- [x]  Implement importing factor level as label

[debug.rds.zip](https://github.com/user-attachments/files/20714955/debug.rds.zip)

shun2wang committed:
initial import of rdata

shun2wang committed:
remove qmlls

some cleaning

EJWagenmakers commented:
Awesome!!

shun2wang committed:
add librdata to github workflow

JorisGoosen commented:
Looks good at first glance!
Nice ^^

Ill have  look next week, and this might even make make some datawrangling easier!

shun2wang commented:
Is there a way to suppress the compiling warning of R package `RcppEigen`? it annoying and maks build time too long. @RensDofferhoff

shun2wang committed:
fix find librdata library on Linux (workflow)

JorisGoosen commented:
> Is there a way to suppress the compiling warning of R package `RcppEigen`? it annoying and maks build time too long. @RensDofferhoff

Its a good idea to add this yeah, I like seeing the output of rpkgs being compiled in case there is an error (which is why I turned it on) but rcppeigen is ridiculous

shun2wang commented:
> > Is there a way to suppress the compiling warning of R package `RcppEigen`? it annoying and maks build time too long. @RensDofferhoff
> 
> Its a good idea to add this yeah, I like seeing the output of rpkgs being compiled in case there is an error (which is why I turned it on) but rcppeigen is ridiculous

Yes, I also support turning on some compilation parameters to identify risks.
Seems it will not be "fixed". see: https://github.com/RcppCore/RcppEigen/issues/127

shun2wang committed:
copy header files to keep them up to date

shun2wang commented:
Since we still haven't frozen the code, can this PR be merged into the next version? @boutinb @JorisGoosen

shun2wang committed:
add build docs on windows

JorisGoosen commented:
Well sorry but it still doesnt work on macos right?

And also, im a bit too busy fixing other issues for the release now to help you out with this.

shun2wang commented:
> Well sorry but it still doesnt work on macos right?
> 
> And also, im a bit too busy fixing other issues for the release now to help you out with this.

Yes, it not yet, but my [PR](https://github.com/conan-io/conan-center-index/pull/26024) with librdata on conan was just merged, so it should be very easy to get it work on macOS I think. I'll blind code it so someone can test it.

shun2wang committed:
maybe this is easy on MacOS

shun2wang committed:
fix build

shun2wang commented:
OK now it could build on macOS I guess?

JorisGoosen commented:
Ill give it a try

JorisGoosen commented:
I get 
```
[cmake] CMake Error at Tools/CMake/Libraries.cmake:253 (find_package):
[cmake]   find_package called with invalid argument "0.0.0.cci.20231003"
```
But I do see it on conancenter, maybe the conan I have doesnt like this commit type?

JorisGoosen committed:
fix macos build

JorisGoosen commented:
cmake was complaining, not conan

shun2wang commented:
> cmake was complaining, not conan

yeah,that version numbers not be accepted in cmake, could be 0.0.0. but it's already on Conan-center.

JorisGoosen commented on Tools/CMake/Libraries.cmake:
```diff
@@ -266,7 +287,8 @@ if(WIN32)
     message(CHECK_PASS "found")
-    message(STATUS "  ${RTOOLS_LIBREADSTAT_H}")
+    message(STATUS "Now copy ${RTOOLS_LIBREADSTAT_H} to source directory.")
+    configure_file("${RTOOLS_LIBREADSTAT_H}" "${CMAKE_SOURCE_DIR}/Desktop/data/importers/readstat/readstat.h" COPYONLY)
```
Why did you add this?

shun2wang replied:
As you know nowadays on Windows we need to manually compile readstat and librdata from msys2 that comes with Rtools, the header files were hardcoded as part of the JASP source code previously, but later you might forget to update it or it would be more intuitive to automatically copy it if you need to include it in the source code.
So we copy it from build source dir automatically now by cmake,to reduces the workload of maintaining it.

shun2wang replied:
You will noticed I had deleted "[readstat.h](https://github.com/jasp-stats/jasp-desktop/pull/5841/files#diff-c0d8f99c10584ea56a77bd4b6488e326d5ec3e0bc1b2bf39ba18e305ae353f56)" from this PR. It will be copied back by cmake while build.

JorisGoosen commented on Tools/CMake/Libraries.cmake:
```diff
@@ -292,6 +314,62 @@ if(WIN32)
+    if(EXISTS ${RTOOLS_LIBRDATA_H})
+      message(CHECK_PASS "found")
+      message(STATUS "Now copy ${RTOOLS_LIBRDATA_H} to source directory")
+      configure_file("${RTOOLS_LIBRDATA_H}" "${CMAKE_SOURCE_DIR}/Desktop/data/importers/rdata/rdata.h" COPYONLY)
```
And this?

JorisGoosen committed:
make it compile and work on macos

JorisGoosen committed:
dont crash if name is null

shun2wang commented on Desktop/CMakeLists.txt:
```diff
@@ -170,8 +173,8 @@ target_link_libraries(
     # librdata -----------------------------------
-    ${LIBRDATA_LIBRARIES}
     $<$<PLATFORM_ID:Windows>:${RTOOLS_LIBRDATA_DLL_A}>
+    $<$<NOT:$<PLATFORM_ID:Windows>>:librdata::librdata>
```
Hi @JorisGoosen which not work on Linux, should just only set it for conan on macOS

JorisGoosen replied:
Sure!
Im now first trying building on windows again. If that works ill go to linux

shun2wang commented on Desktop/CMakeLists.txt:
```diff
@@ -118,7 +120,7 @@ target_include_directories(
     ${LIBREADSTAT_INCLUDE_DIRS}
     # librdata
-    ${LIBRDATA_INCLUDE_DIRS}
+    librdata::librdata
```
And also this not work on Linux I think, see: https://github.com/jasp-stats/jasp-desktop/actions/runs/15583646306/job/43884505672?pr=5841

JorisGoosen committed:
make windows function again

JorisGoosen committed:
now it works on mac too

JorisGoosen commented:
>  Implement importing factor level as label (not now)

This sounds rather easy to do (and also very nice to have)

shun2wang commented:
Ooops, Linux build broken then😂: https://github.com/jasp-stats/jasp-desktop/actions/runs/15584725479/job/43888010825?pr=5841#step:10:555

shun2wang commented on Desktop/data/importers/rdata/readrdata.h:
```diff
@@ -19,9 +19,14 @@
-#include <rdata.h>
 #include "utils.h"
 
+#ifdef __unix__
```
Why this necessary? I guess `#include "rdata.h"` will always fine? because we had "copied" rdata.h automatically on Windows.

shun2wang replied:
On MacOS conan will help to find headerfile, on Linux it will find in system include path. only Windows is specially to "copy" that header from Rtools.

JorisGoosen commented:
So shun, it turns out I only have access to a linux machine at the earliest tomorrow.
You are on linux right?
Perhaps you can fix it again for linux.

In the meantime Ill add the labels from factors.

shun2wang commented:
> So shun, it turns out I only have access to a linux machine at the earliest tomorrow.
> You are on linux right?
> Perhaps you can fix it again for linux.
> 
> In the meantime Ill add the labels from factors.

sure, thanks a lot.

shun2wang committed:
fix build on Linux

shun2wang commented:
Now both Linux and Windows are build succeed.

JorisGoosen committed:
Use ImportDataSet and ImportColumn directly to read the data into

uses rdata type for columntype
factors now also get read, as ordinal

JorisGoosen reviewed (commented):
This now does everything it needs to do right?

JorisGoosen commented:
I rewrote some stuff to store data directly in `RDataImportColumn` and renamed your `RDataReader` to `RDataSetImporter` and subclassed it from `ImportDataSet`.

I also added support for factors.
Maybe you wanna give it a go and see if it also works for you?

Ive uploaded an rds of the debug dataset (including mangled encoded columnames :p)

JorisGoosen commented:
@RensDofferhoff I assume this means we need to add a small segment to the flatpak json

shun2wang committed:
better handling R missing values

JorisGoosen commented on Desktop/data/importers/rdata/rdataimportdataset.cpp:
```diff
@@ -108,7 +113,7 @@ int RDataImportDataSet::columnHandler(const std::string & name, rdata_type_t typ
 		int32_t *values = static_cast<int32_t *>(data);
 		for (size_t i = 0; i < count; i++)
-			_curCol->addValue(std::to_string(values[i]));
+			_curCol->addValue(values[i] == NA_INTEGER ? "" : std::to_string(values[i]));
```
Ah yeah nice

shun2wang commented:
> I rewrote some stuff to store data directly in `RDataImportColumn` and renamed your `RDataReader` to `RDataSetImporter` and subclassed it from `ImportDataSet`.
> 
> I also added support for factors.
> Maybe you wanna give it a go and see if it also works for you?
> 
> Ive uploaded an rds of the debug dataset (including mangled encoded columnames :p)

thanks for your work!

now minor issue is empty colnames will not be encoded correctly?

JorisGoosen commented:
> thanks for your work!

You too! ^^

> now minor issue is empty colnames will not be encoded correctly?

I only tested with a colnameless dataframe once and saw a `V1` column I thought?
Just like the standard JASP loading?

Im afraid I dont have anymore time today though...
But @boutinb is going to test and review this as well, as Ive now changed so much it probably needs another reviewer too :p
Maybe he can doublecheck?

shun2wang commented:
>I only tested with a colnameless dataframe once and saw a V1 column I thought?
Just like the standard JASP loading?

I mean like this, "col3" is empty name in .Rdata and imported as V2, maybe should be `index + 1` then V3 is correct?

![image](https://github.com/user-attachments/assets/1500103e-80e5-4a24-af23-1020cf576116)

JorisGoosen commented:
Does the same happen with csvs (cause its the same code as far as I could see)?
Otherwise, ill (or someone'll) have to have a look.

shun2wang commented:
> Does the same happen with csvs (cause its the same code as far as I could see)? Otherwise, ill (or someone'll) have to have a look.

No, csv is not same code to handle it: https://github.com/jasp-stats/jasp-desktop/blob/ec219606acdb9ddc83c78e4ea58bd7fa4cb17f3d/Desktop/data/importers/csvimporter.cpp#L43-L49

JorisGoosen commented:
Then you indeed just need `curCol+1` t https://github.com/jasp-stats/jasp-desktop/blob/ec219606acdb9ddc83c78e4ea58bd7fa4cb17f3d/Desktop/data/importers/importdataset.cpp#L110

shun2wang commented:
> Then you indeed just need `curCol+1` t
> 
> https://github.com/jasp-stats/jasp-desktop/blob/ec219606acdb9ddc83c78e4ea58bd7fa4cb17f3d/Desktop/data/importers/importdataset.cpp#L110

Ah, hahaha I'm so stupid I was even not noticed it😂

shun2wang committed:
fix empty cols names encode

shun2wang commented:
Now everything is done. @boutinb Do you want to have a look?

boutinb commented:
I have tested it, it works fine, but I get an extra column ".Random.seed" with quite strange numbers, and much too many labels for this column...

boutinb commented:
[Simple.RData.zip](https://github.com/user-attachments/files/20727774/Simple.RData.zip)
This is the RData file I used.

JorisGoosen commented:
Maybe you have a weird dataset?
In any case, I load every level of a factor as a label and thus there could be more labels than values for sure.

shun2wang commented:
> I have tested it, it works fine, but I get an extra column ".Random.seed" with quite strange numbers, and much too many labels for this column...

This is because another internal "table" was read as cols, in debug log you can see a table named ".Random.seed", R didn't read it explicit...

But this is a behavior of librdata even it's python bindings (https://github.com/ofajardo/pyreadr) imported same _random.seed cols. I made an issue to track it. https://github.com/WizardMac/librdata/issues/50

shun2wang commented:
This should also into testing version or not? @boutinb
